### PR TITLE
Add strike and underline span formatting

### PR DESCRIPTION
**Pull request by roman-danilov**

### CODE REVIEW

File: html2docx.py [hunk 1]
```diff
@@ -10,6 +10,16 @@
 WHITESPACE_RE = re.compile(r"\s+")
 
 
+def get_attrs(attrs: List[Tuple[str, Optional[str]]]) -> List[str]:
```
Comment (jdufresne) [inline html2docx.py]: I find this function name very terse. It is named "get_attrs", accepts a argument "attrs" and returns a variable "args". I think it could be a bit more descriptive or have a docstring to describe its purpose.

File: html2docx.py [hunk 2]
```diff
@@ -10,6 +10,16 @@
...
+    args = []
+    for attr in attrs:
+        if attr == ("style", "text-decoration: underline;"):
+            args.append("underline")
+        elif attr == ("style", "text-decoration: line-through;"):
```
Comment (jdufresne) [inline html2docx.py]: The style attribute may not always have exactly this whitespace or formatting. It could also contain multiple inline styles. I think we'll need to some basic CSS parsing. Perhaps an existing library already does this. More tests for these edge cases would be nice.

File: html2docx.py [hunk 3]
```diff
@@ -25,21 +35,24 @@ def _reset(self) -> None:
...
-    def init_run(self, attrs: List[str]) -> None:
-        self.r = None
+    def init_run(self, attrs: Optional[List[str]]) -> None:
         self.attrs.append(attrs)
+        if attrs is not None:
+            self.r = None
 
     def finish_run(self) -> None:
+        attrs = self.attrs and self.attrs[-1]
         self.attrs = self.attrs[:-1]
```
Comment (jdufresne) [inline html2docx.py]: These two lines can be combined to:

```
attrs = self.attrs.pop()
```

File: html2docx.py [hunk 4]
```diff
@@ -77,6 +91,9 @@ def handle_starttag(self, tag: str, attrs: List[Tuple[str, Optional[str]]]) -> N
             self.add_list_style("List Number")
         elif tag == "pre":
             self.pre = True
+        elif tag == "span":
+            span_attrs = get_attrs(attrs) or None
```
Comment (jdufresne) [inline html2docx.py]: I think we could avoid this `or None`

```diff
diff --git a/html2docx.py b/html2docx.py
index debad3b..fad07e5 100644
--- a/html2docx.py
+++ b/html2docx.py
@@ -35,7 +35,7 @@ class HTML2Docx(HTMLParser):
 
         # Formatting options
         self.pre = False
-        self.attrs: List[Optional[List[str]]] = []
+        self.attrs: List[List[str]] = []
         self.collapse_space = True
 
     def finish_p(self) -> None:
@@ -43,9 +43,9 @@ class HTML2Docx(HTMLParser):
             self.r.text = self.r.text.rstrip()
         self._reset()
 
-    def init_run(self, attrs: Optional[List[str]]) -> None:
+    def init_run(self, attrs: List[str]) -> None:
         self.attrs.append(attrs)
-        if attrs is not None:
+        if attrs:
             self.r = None
 
     def finish_run(self) -> None:
@@ -92,7 +92,7 @@ class HTML2Docx(HTMLParser):
         elif tag == "pre":
             self.pre = True
         elif tag == "span":
-            span_attrs = get_attrs(attrs) or None
+            span_attrs = get_attrs(attrs)
             self.init_run(span_attrs)
         elif tag == "sub":
             self.init_run(["subscript"])
```

**Comment by jdufresne** [comment]
What does "delete" represent in the test name? If it means "strikethrough", perhaps that would better name.

**Comment by roman-danilov** [comment]
- `get_attrs` doc string
- css processing utility
- operations short cut `attrs = self.attrs and self.attrs[-1]`
- change test file name

File: setup.py [hunk 5]
```diff
@@ -32,5 +32,5 @@ def readall(path):
         "Topic :: Utilities",
     ],
     python_requires=">=3.6",
-    install_requires=["python-docx"],
+    install_requires=["python-docx", "cssutils"],
```
Comment (jdufresne) [inline setup.py]: This package hasn't had a commit or release since 2017-04-14. It is only tested against Python 3.5. It looks unmaintained. Could we pick a a better maintained library?

Would https://github.com/Kozea/tinycss2 be suitable? I'm also open to other suggestions if you have them.

File: html2docx.py [hunk 6]
```diff
@@ -3,13 +3,40 @@
...
 WHITESPACE_RE = re.compile(r"\s+")
 
 
+def attr_to_css(attr: Tuple[str, Optional[str]]) -> Optional[CSSStyleDeclaration]:
+    if attr[0] == "style":
+        return parseStyle(attr[1])
+    return None
+
+
+def get_attrs(attrs: List[Tuple[str, Optional[str]]]) -> List[str]:
```
Comment (jdufresne) [inline html2docx.py]: WDYT about: `html_attrs_to_font_style` for a name?

File: html2docx.py [hunk 7]
```diff
@@ -25,21 +52,23 @@ def _reset(self) -> None:
...
-        self.r = None
+    def init_run(self, attrs: Optional[List[str]]) -> None:
         self.attrs.append(attrs)
+        if attrs is not None:
+            self.r = None
 
     def finish_run(self) -> None:
-        self.attrs = self.attrs[:-1]
-        self.r = None
+        attrs = self.attrs and self.attrs.pop()
```
Comment (jdufresne) [inline html2docx.py]:
```suggestion
        attrs = self.attrs.pop()
```

File: html2docx.py [hunk 8]
```diff
@@ -3,13 +3,40 @@
...
+
+    :param attrs:
+        a list of attribute name plus attribute value tuples.
+    :returns:
+        a list of style names to be applied to Run Font property
+    """
+    args = []
+    for attr in attrs:
+        css = attr_to_css(attr)
+        if css:
```
Comment (jdufresne) [inline html2docx.py]: Removing this guard does not cause any test failures. When is `css` `None`? A test should be added.

**Comment by roman-danilov** [comment]
- Use tinycss2 instead of cssutils
- Rename test file
- function name `get_attrs` -> `html_attrs_to_font_style`

File: html2docx.py [hunk 9]
```diff
@@ -6,10 +6,39 @@
...
 from docx.text.paragraph import Paragraph
 from docx.text.run import Run
+from tinycss2 import parse_one_declaration
+from tinycss2.ast import IdentToken
 
 WHITESPACE_RE = re.compile(r"\s+")
 
 
+def attr_to_css(attr: Tuple[str, Optional[str]]) -> Optional[Tuple[str, str]]:
+    if attr[0] == "style":
```
Comment (jdufresne) [inline html2docx.py]: I think tuple unpacking would aid readability. 0 and 1 are a bit opaque. Something like:

```diff
diff --git a/html2docx.py b/html2docx.py
index 9d20679..919d6d5 100644
--- a/html2docx.py
+++ b/html2docx.py
@@ -13,8 +13,9 @@ WHITESPACE_RE = re.compile(r"\s+")


 def attr_to_css(attr: Tuple[str, Optional[str]]) -> Optional[Tuple[str, str]]:
-    if attr[0] == "style":
-        declaration = parse_one_declaration(attr[1])
+    name, value = attr
+    if name == "style":
+        declaration = parse_one_declaration(value)
         for value in declaration.value:
             if isinstance(value, IdentToken):
                 return declaration.lower_name, value.lower_value
```

File: html2docx.py [hunk 10]
```diff
@@ -6,10 +6,39 @@
...
+        declaration = parse_one_declaration(attr[1])
+        for value in declaration.value:
+            if isinstance(value, IdentToken):
+                return declaration.lower_name, value.lower_value
```
Comment (jdufresne) [inline html2docx.py]: This early return assumes there is only a single style. Please include a test with multiple styles. For example:

```html
<span style="font-size:9pt;text-decoration:underline;">Lorem ipsum</span>
```

File: html2docx.py [hunk 11]
```diff
@@ -6,10 +6,39 @@
...
+
+def html_attrs_to_font_style(attrs: List[Tuple[str, Optional[str]]]) -> List[str]:
+    """Return Font style names based on tag style attributes
+
+    :param attrs:
+        a list of attribute name plus attribute value tuples.
+    :returns:
+        a list of style names to be applied to Run Font property
+    """
+    args = []
```
Comment (jdufresne) [inline html2docx.py]: Maybe `style` would signal the intent better. WDYT?

**Comment by roman-danilov** [comment]
- Re-factor `attr_to_css`
- Change variable name

File: html2docx.py (hunk 10 above)
Comment (roman-danilov) [reply html2docx.py]: No support for other inline element styles was designed. Only `text-decoration`. Even a case with e. g. subscript and underline will not be handled (support for this is easy to add, if you think it is required).
I added test for two styles combined.

**Comment by jdufresne** [comment]
The following diff does not cause any test failures for me. I think there is something wrong with the strike through tests:

```diff
diff --git a/tests/data/underline-strikethrough.json b/tests/data/underline-strikethrough.json
index 30a32d6..b82c930 100644
--- a/tests/data/underline-strikethrough.json
+++ b/tests/data/underline-strikethrough.json
@@ -7,7 +7,6 @@
             },
             {
                 "text": "very",
-                "strike": true,
                 "underline": true
             },
             {
@@ -24,8 +23,7 @@
         "runs": [
             {
                 "text": "both underline and strike-through",
-                "underline": true,
-                "strike": true
+                "underline": true
             }
         ]
     }
```

**Comment by roman-danilov** [comment]
Thanks, good catch.

**Comment by jdufresne** [comment]
The code makes some assumptions about the exact layout of the CSS that I think we could do away with to be more robust. For example, I expect the following test to pass, but it does not. 

```diff
diff --git a/tests/data/underline-strikethrough.html b/tests/data/underline-strikethrough.html
index 3a3bb46..a84024f 100644
--- a/tests/data/underline-strikethrough.html
+++ b/tests/data/underline-strikethrough.html
@@ -1,2 +1,3 @@
 <p>This is a <span style="text-decoration: underline;"><span style="text-decoration: line-through;">very</span> important</span> issue.</p>
 <p><span style="text-decoration: underline;"><span style="text-decoration: line-through;">both underline and strike-through</span></span></p>
+<p>Lorem <span style="text-decoration:underline line-through;">ipsum</span>.</p>
diff --git a/tests/data/underline-strikethrough.json b/tests/data/underline-strikethrough.json
index 30a32d6..576e09a 100644
--- a/tests/data/underline-strikethrough.json
+++ b/tests/data/underline-strikethrough.json
@@ -28,5 +28,21 @@
                 "strike": true
             }
         ]
+    },
+    {
+        "text": "Lorem ipsum.",
+        "runs": [
+            {
+                "text": "Lorem "
+            },
+            {
+                "text": "ipsum",
+                "underline": true,
+                "strike": true
+            },
+            {
+                "text": "."
+            }
+        ]
     }
 ]
```

**Comment by jdufresne** [comment]
WDYT about the following simplification:

- It makes fewer types optional lending to stricter types
- It combines adjacent runs with the same font attributes
- Passes the full test matrix for me

```diff
diff --git a/html2docx.py b/html2docx.py
index 84b6567..a5d4a6c 100644
--- a/html2docx.py
+++ b/html2docx.py
@@ -55,7 +55,7 @@ class HTML2Docx(HTMLParser):
 
         # Formatting options
         self.pre = False
-        self.attrs: List[Optional[List[str]]] = []
+        self.attrs: List[List[str]] = []
         self.collapse_space = True
 
     def finish_p(self) -> None:
@@ -63,14 +63,14 @@ class HTML2Docx(HTMLParser):
             self.r.text = self.r.text.rstrip()
         self._reset()
 
-    def init_run(self, attrs: Optional[List[str]]) -> None:
+    def init_run(self, attrs: List[str]) -> None:
         self.attrs.append(attrs)
-        if attrs is not None:
+        if attrs:
             self.r = None
 
     def finish_run(self) -> None:
         attrs = self.attrs.pop()
-        if attrs is not None:
+        if attrs:
             self.r = None
 
     def add_text(self, data: str) -> None:
@@ -80,9 +80,8 @@ class HTML2Docx(HTMLParser):
         if self.r is None:
             self.r = self.p.add_run()
             for attrs in self.attrs:
-                if attrs:
-                    for run_style in attrs:
-                        setattr(self.r.font, run_style, True)
+                for run_style in attrs:
+                    setattr(self.r.font, run_style, True)
         self.r.add_text(data)
 
     def add_list_style(self, name: str) -> None:
@@ -111,7 +110,7 @@ class HTML2Docx(HTMLParser):
         elif tag == "pre":
             self.pre = True
         elif tag == "span":
-            span_attrs = html_attrs_to_font_style(attrs) or None
+            span_attrs = html_attrs_to_font_style(attrs)
             self.init_run(span_attrs)
         elif tag == "sub":
             self.init_run(["subscript"])
diff --git a/tests/data/inner-link.json b/tests/data/inner-link.json
index 4d84f33..a5d4e09 100644
--- a/tests/data/inner-link.json
+++ b/tests/data/inner-link.json
@@ -6,11 +6,7 @@
                 "text": "The "
             },
             {
-                "text": "localhost http://localhost",
-                "bold": true
-            },
-            {
-                "text": " link",
+                "text": "localhost http://localhost link",
                 "bold": true
             },
             {
diff --git a/tests/data/link-with-space.json b/tests/data/link-with-space.json
index a5d021a..70d3e6d 100644
--- a/tests/data/link-with-space.json
+++ b/tests/data/link-with-space.json
@@ -3,10 +3,7 @@
         "text": "Localhost http://localhost link.",
         "runs": [
             {
-                "text": "Localhost http://localhost "
-            },
-            {
-                "text": "link."
+                "text": "Localhost http://localhost link."
             }
         ]
     }
```

**Comment by roman-danilov** [comment]
Change style attribute handling.
